doc/amoc_throttle.md

```diff
@@ -0,0 +1,180 @@
+Amoc throttle makes sure that the given `Rate` per `Interval` is maintained on a constant level.
+It prevents bursts of executions which could blurry the results, as they technically produce a desired rate in a given interval.
+Because of that, it may happen that the actual `Rate` would be slightly below the demanded rate. However, it will never be exceeded.
+
+## API
+
+Functions provided by Amoc throttle:
+```
+start(Name :: name(), Rate :: pos_integer()[, Interval :: non_neg_integer(), NoOfProcesses :: pos_integer()) -> ok | {error, any()}.
+```
+Starts the throttle mechanism for a given `Name` with a given `Rate`.
+The optional arguments are an `Interval` (default is one minute) and a ` NoOfProcesses` (default is 10).
+`Name` is needed to identify the rate as a single test can have different rates for different tasks.
+`Interval` is given in milliseconds and can be changed to a different value for convenience or higher granularity.
+It also accepts a special value of `0` which limits the number of parallel executions associated with `Name` to `Rate`.
+
+```
+send(Name :: name()[, PID :: pid()], Msg :: any()) -> ok | {error, any()}.
+```
+Sends a given message `Msg` either to `self()` or a given (optional) `PID` when the rate for `Name` allows for that.
+May be used to schedule tasks.
+
+```
+send_and_wait(Name :: name(), Msg :: any()) -> ok | {error, any()}.
+```
+Sends and receives the given message `Msg`.
+Can be used to halt execution if we want a process to be idle when waiting for rate increase or other processes finishing their tasks.
+```
+run(Name :: name(), Fn :: fun(()-> any())) -> ok | {error, any()}.
+```
+Executes a given function `Fn` when it does not exceed the rate for `Name`.
+`Fn` is executed in the context of a new process spawned on the same node on which the process executing `run/2` runs, so a call to `run/2` is non-blocking.
+This function is used internally by both `send` and `send_and_wait/2` functions,
+so all those actions will be limited to the same rate when called with the same `Name`.
+
+```
+pause(Name :: name()) -> ok | {error, any()}.
+```
+Pauses executions for the given `Name` as if `Rate` was set to `0`.
+Does not stop the scheduled rate changes.
+
+```
+resume(Name :: name()) -> ok | {error, any()}.
+```
+Resumes the executions for the given `Name`, so the `Rate` and `Interval` are reset to their original values given before the pause.
+```
+change_rate(name(), pos_integer(), non_neg_integer()) -> ok | {error, any()}.
+```
+Sets `Rate` and `Interval` for `Name` according to the given values.
+Can change whether Amoc throttle limits `Name` to parallel executions or to `Rate` per `Interval`, according to the given `Interval` value.
+```
+change_rate_gradually(Name :: name(), From :: pos_integer(), To :: pos_integer(), RateInterval :: non_neg_integer(), StepInterval :: pos_integer(), NoOfSteps :: pos_integer()) -> ok | {error, any()}.
```

Comment:
```suggestion
change_rate_gradually(Name, From, To, RateInterval, StepInterval, NoOfSteps)) -> R when
                      Name :: name(),
                      From :: pos_integer(),
                      To :: pos_integer(),
                      RateInterval :: non_neg_integer(),
                      StepInterval :: pos_integer(),
                      NoOfSteps :: pos_integer(),
                      R :: ok | {error, any()}.
```

I think this style is way more readable for specs/signatures this long.
The original line was 199 chars!

Reply:
True 👍